modules/shared_data/src/main/scala/org/elpaca_metagraph/shared_data/Utils.scala

```diff
@@ -100,9 +102,16 @@ object Utils {
         password.value.value.toCharArray
       )
 
-  def randomString(length: Int): String = {
+  def randomString[F[_]: Async](input: String, length: Int): F[String] = {
+    val hash = Hash.fromBytes(input.getBytes("utf-8")).toString
+    val seed = BigInt(hash.take(16), 16).toLong
+
     val chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
-    Random.alphanumeric.filter(chars.contains(_)).take(length).mkString
+
+    Random.scalaUtilRandomSeedLong[F](seed).flatMap { random =>
+      (1 to length).toList.traverse(_ => random.nextIntBounded(chars.length))
```

Comment:
Look at a function "nextAlphaNumeric". You don't need to do it manually. Also, there is "nextString(lenght)" so you can check both fns :) The rest in the PR is ok 👍